src/encoder/encodermp3.cpp

```diff
@@ -134,9 +149,53 @@ void EncoderMp3::flush() {
         numBytes = lame_get_lametag_frame(
                 m_lameFlags, m_bufferOut, m_bufferOutSize);
     }
-    // Write the lame/xing header.
+
+    TagLib::ID3v2::Tag id3Tag;
+
+    // Ideally, we should shift the file content forward to insert the header
+    // (ID3v2 & Xing frame) dynamically, but `EncoderCallback` doesn't support
+    // that so we use the static header padding and truncate the tracklist if
+    // too long
+    if (!m_metaDataTitle.isEmpty()) {
+        id3Tag.setTitle(QStringToTString(m_metaDataTitle));
+    }
+    if (!m_metaDataArtist.isEmpty()) {
+        id3Tag.setArtist(QStringToTString(m_metaDataArtist));
+    }
+    if (!m_metaDataAlbum.isEmpty()) {
+        id3Tag.setAlbum(QStringToTString(m_metaDataAlbum));
+    }
+    qsizetype tracklistMaxByteSize = kHeaderPadding -
+            id3Tag.render().size();
+    DEBUG_ASSERT(tracklistMaxByteSize > 0);
+    TagLib::String trackList = QStringToTString(getTrackList().join("\n"));
```

Comment:
Instead of join() you can use a while loop and add tracks until the size is reached and the remove the last.